sdk/vppagent/cross_connect.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/networkservicemesh/networkservicemesh/sdk/endpoint"
+	"github.com/sirupsen/logrus"
+)
+
+// XConnect is a VPP Agent Cross Connect composite
+type XConnect struct {
+	endpoint.BaseCompositeEndpoint
+	Workspace   string
+	Connections map[string]*ConnectionData
+}
+
+// Request implements the request handler
+func (xc *XConnect) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+
+	if xc.GetNext() == nil {
+		logrus.Fatal("The VPP Agent Cross Connect composite requires that there is Next set")
+	}
+
+	incomingConnection, err := xc.GetNext().Request(ctx, request)
+	if err != nil {
+		logrus.Errorf("Next request failed: %v", err)
+		return nil, err
+	}
+
+	opaque := xc.GetNext().GetOpaque(incomingConnection)
+	if opaque == nil {
+		err := fmt.Errorf("received empty data from Next")
+		logrus.Errorf("Unable to find connection data: %v", err)
+		return nil, err
+	}
+	connectionData := opaque.(*ConnectionData)
+
+	if connectionData.DstName == "" {
+		err := fmt.Errorf("found empty destination name")
+		logrus.Errorf("Invalid connection data: %v", err)
+		return nil, err
+	}
+	if connectionData.SrcName == "" {
+		err := fmt.Errorf("found empty source name")
+		logrus.Errorf("Invalid connection data: %v", err)
+		return nil, err
+	}
+
+	connectionData.DataChange = xc.appendDataChange(connectionData.DataChange, connectionData.SrcName, connectionData.DstName)
+
+	xc.Connections[incomingConnection.GetId()] = connectionData
+	return incomingConnection, nil
+}
+
+// Close implements the close handler
+func (xc *XConnect) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
+	if xc.GetNext() != nil {
+		return xc.GetNext().Close(ctx, connection)
+	}
+	return &empty.Empty{}, nil
+}
+
+// GetOpaque will return the corresponding connection data
+func (xc *XConnect) GetOpaque(incoming interface{}) interface{} {
+	incomingConnection := incoming.(*connection.Connection)
+	if connectionData, ok := xc.Connections[incomingConnection.GetId()]; ok {
+		return connectionData
+	}
+	logrus.Errorf("GetOpaque outgoing not found for %v", incomingConnection)
+	return nil
+}
+
+// Name returns the composite name
+func (xc *XConnect) Name() string {
+	return "cross-connect"
+}
+
+// NewXConnect creates a XConnect
+func NewXConnect(configuration *common.NSConfiguration) *XConnect {
+	// ensure the env variables are processed
+	if configuration == nil {
+		configuration = &common.NSConfiguration{}
+	}
+	configuration.CompleteNSConfiguration()
+
+	return &XConnect{
+		Workspace:   configuration.Workspace,
+		Connections: map[string]*ConnectionData{},
+	}
+}
+
+func (xc *XConnect) appendDataChange(rv *configurator.Config, srcName, dstName string) *configurator.Config {
+	if rv == nil {
+		rv = &configurator.Config{}
+	}
+	if rv.VppConfig == nil {
+		rv.VppConfig = &vpp.ConfigData{}
+	}
+
+	rv.VppConfig.XconnectPairs = append(rv.VppConfig.XconnectPairs,
```

Comment:
TODO for all composites: always return a copy of DataChange object, do not mutate opaque data of the next composite

Reply:
Hmm interesting. Return it where? Are you talkoing about a different method?

Reply:
Yes, not return, I'm about storing in opaque. I mean we send DataChange object through the chain and most likely we want to have snapshots of this object in each composite. Currently I mutate one DataChange instance in all composites. I think it will be good to copy DataChange object in each composite and mutate only this copy, do not affect other composites opaque.

Reply:
Done